src/celeritas/track/TrackFunctors.hh

```diff
@@ -0,0 +1,120 @@
+//---------------------------------------------------------------------------//
+//! \file celeritas/track/TrackFunctors.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/sys/ThreadId.hh"
+#include "celeritas/Types.hh"
+#include "celeritas/field/UniformFieldData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+// CONDITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Condition for ConditionalTrackExecutor for active, non-errored tracks.
+ */
+struct AppliesValid
+{
+    template<class T>
+    CELER_FUNCTION bool operator()(T const& track) const
+    {
+        return is_track_valid(track.sim().status());
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Apply only to tracks with the given post-step action ID.
+ */
+struct IsStepActionEqual
+{
+    ActionId action;
+
+    template<class T>
+    CELER_FUNCTION bool operator()(T const& track) const
+    {
+        return track.sim().post_step_action() == this->action;
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Apply only to tracks with the given along-step action ID.
+ */
+struct IsAlongStepActionEqual
+{
+    ActionId action;
+
+    template<class T>
+    CELER_FUNCTION bool operator()(T const& track) const
+    {
+        CELER_EXPECT(AppliesValid{}(track)
+                     == static_cast<bool>(track.sim().along_step_action()));
+        return track.sim().along_step_action() == this->action;
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Apply only to tracks in a volume with a field.
+ */
+struct IsInUniformField
+{
+    NativeCRef<UniformFieldParamsData> field;
+
+    template<class T>
+    CELER_FUNCTION bool operator()(T const& track) const
+    {
+        if (field.has_field.empty())
+        {
+            // Field is present in all volumes
+            return true;
+        }
+        auto vol = track.geometry().volume_id();
+        CELER_ASSERT(vol < field.has_field.size());
+        return field.has_field[vol];
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Apply to tracks in the uniform along-step action in volumes with field.
+ */
+struct IsAlongStepUniformField
+{
+    ActionId action;
+    NativeCRef<UniformFieldParamsData> field;
+
+    template<class T>
+    CELER_FUNCTION bool operator()(T const& track) const
+    {
+        return IsAlongStepActionEqual{action}(track)
+               && IsInUniformField{field}(track);
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Apply to tracks in the uniform along-step action in volumes without field.
+ */
+struct IsAlongStepLinear
```

Comment:
You're not going to define a variadic template functor that returns "all of" the `operator()` of its class arguments? 🙃  (j/k let's not)